In [1]:
# imports
import pandas as pd
import numpy as np

import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

from plotly.offline import iplot, init_notebook_mode
init_notebook_mode(connected=True)


# load dataset
df = pd.read_csv('/kaggle/input/gfg-pat/finalPAT.csv', encoding='latin-1')


# first 5 records
df.head()

Name Age Net_Worth           Income Source    Citizenship
0       Paul Saville   66    $1.6 B            Real Estate   United States
1         Georg Haub   60    $2.2 B       Fashion & Retail         Germany
2        Arvind Tiku   52    $2.2 B                 Energy           India
3         Bill Gross   77    $1.6 B  Finance & Investments   United States
4  Hartono Kweefanus   72    $1.2 B        Food & Beverage       Indonesia

# Q1: Which is the 8th country that has produced maximum number of billionaires?

In [2]:
# group 'Citizenship' column, count the size of each country and take the top 10

top10 = df.groupby('Citizenship').size().reset_index(name='count').nlargest(10, 'count')

In [3]:
# slice 8th position

top10[7:8]

Citizenship  count
11      Brazil     62

### **Brazil** is the 8th country that has produced maximum number of *62* billionaires.

# Q2: How many Duplicate Values are present in the Dataframe
* Find Duplicate using Name

In [4]:
# take 'Name' column and check for duplicates

df['Name'].duplicated().sum()

2

### There are **2** duplicate *Name* values are present in the Dataframe.

# Q3: Which is the industry with the highest and Lowest average net worth for billionaires in the Forbes dataset?
* Industry here refers to Income Source 
* Check for type mismatch errors before finding mean 
* You have to find the mean of all the similar Income Sources and find the highest among them and then lowest among them

In [5]:
# The 'Net_Worth' column consists of $(symbol for US dollars) and B(unit for Billion) notations
# remove '$' and 'B' from each values of 'Net_Worth' column and change its type to 'float'

df['Net_Worth'] = df['Net_Worth'].str.strip('$ B').astype(float)

In [6]:
# check for the values of 'Income Source' column

df['Income Source'].unique()

array(['Real Estate ', 'Fashion & Retail ', 'Energy ',
       'Finance & Investments ', 'Food & Beverage ', 'Technology ',
       'Diversified ', 'Media & Entertainment ', 'Manufacturing ',
       'Healthcare ', 'Telecom ', 'Metals & Mining ', 'Automotive ',
       'Service ', 'chemicals, Self Made', 'Logistics ', 'biscuits',
       'Sports ', 'edtech, Self Made', 'Gambling & Casinos ',
       'Construction & Engineering ', 'steel', 'medical diagnostics',
       'Bangalore, India', 'software services, Self Made',
       'financial technology, Self Made', 'tractors',
       'mining, metals, Self Made', 'Mumbai, India', 'Delhi, India',
       'consumer goods, Self Made', 'construction, Self Made',
       'real estate', 'software services', 'paints', 'diversified',
       'chemicals', 'motorcycles', 'Chennai, India',
       'plastic pipes, Self Made', 'consumer goods', 'pharmaceuticals',
       'pharmaceuticals, Self Made', 'Ireland', 'healthcare, Self Made',
       'cement', 'Ahmedabad, 

In [7]:
# place India and Self Made are not considered as Industries

self_made = df[df['Income Source'].str.contains('Self')]

place_india = df[df['Income Source'].str.contains('India')]

In [8]:
# remove invalid data for this question

df.drop(self_made.index, inplace=True)
df.drop(place_india.index, inplace=True)

In [9]:
# group 'Income Source' column and take 'Net_Worth' column to calculate mean of each industry

industry_networth_avg = df.groupby('Income Source')['Net_Worth'].mean().reset_index(name='mean')

In [10]:
# industry with highest average networth

industry_networth_avg.nlargest(1, 'mean')

Income Source  mean
43      vaccines  21.5

In [11]:
# industry with lowest average networth

industry_networth_avg.nsmallest(1, 'mean')

Income Source  mean
31  medical diagnostics   1.5

### The industry with the highest and Lowest average net worth are **Vaccines** *(21.5B dollars)* and **Medical Diagnostics** *($1.5 B)* respectively.

# Data Cleaning

In [12]:
# take a look into invalid data

place_india

Name  \
150                                Dilip & Anand Surana   
418   Girdhari Lal Bawri, Rajendra Agarwal and Banwa...   
447                              Vinod & Anil Rai Gupta   
542                                Amalgamations family   
828                                    Murugappa Family   
937                                Sudhir & Samir Mehta   
1016                    Ramesh Kumar and Mukand Lal Dua   
1162                                      Godrej family   
1195                                       Bajaj Family   
1379                                       Singh family   
1599                               Kapil & Rahul Bhatia   
1677                       Chandrakant & Rajendra Gogri   
1772                                     Muthoot family   
1845                                       Gupta family   
1908                                Rakesh Jhunjhunwala   
1911             Kuldip Singh & Gurbachan Singh Dhingra   
1912                                      Burman family   
2310                                       Reddy Family   
2471                           Nithin and Nikhil Kamath   
2692                                        Abhay Vakil   

                                Age  Net_Worth     Income Source  \
150                 pharmaceuticals       2.25  Bangalore, India   
418      pharmaceuticals, Self Made       6.50     Mumbai, India   
447            electrical equipment       7.60      Delhi, India   
542                        tractors       2.89    Chennai, India   
828                     diversified       4.70    Chennai, India   
937          pharmaceuticals, power       7.10  Ahmedabad, India   
1016                       footwear       2.85      Delhi, India   
1162    consumer goods, real estate      15.20     Mumbai, India   
1195                    diversified      14.40       Pune, India   
1379                pharmaceuticals       5.65     Mumbai, India   
1599            airlines, Self Made       5.00      Delhi, India   
1677           chemicals, Self Made       2.10     Mumbai, India   
1772             financial services       6.40      Delhi, India   
1845                pharmaceuticals       2.82     Mumbai, India   
1908         investments, Self Made       5.50             India   
1911              paints, Self Made       9.40      Delhi, India   
1912                 consumer goods      11.80      Delhi, India   
2310                pharmaceuticals       3.01  Hyderabad, India   
2471  financial services, Self Made       2.59  Bangalore, India   
2692                         paints       6.25             India   

                                            Citizenship  
150                                                 NaN  
418                                                 NaN  
447                                                 NaN  
542                                                 NaN  
828                                                 NaN  
937        Bachelor of Arts/Science, Gujarat University  
1016                                                NaN  
1162                                                NaN  
1195                                                NaN  
1379                                                NaN  
1599  Bachelor of Science in Engineering, University...  
1677  Bachelor of Science in Engineering, University...  
1772                                                NaN  
1845                                                NaN  
1908  Chartered Accountant, The Institute of Charter...  
1911                                            Married  
1912                                                NaN  
2310                                                NaN  
2471                                                NaN  
2692  Bachelor of Arts/Science, Syracuse University;...

In [13]:
# column values are misplaced

india = place_india.copy()

india['Income Source'] = india['Age']

india['Citizenship'] = 'India'

india['Age'] = np.nan

In [14]:
india.head()

Name  Age  Net_Worth  \
150                               Dilip & Anand Surana  NaN       2.25   
418  Girdhari Lal Bawri, Rajendra Agarwal and Banwa...  NaN       6.50   
447                             Vinod & Anil Rai Gupta  NaN       7.60   
542                               Amalgamations family  NaN       2.89   
828                                   Murugappa Family  NaN       4.70   

                  Income Source Citizenship  
150             pharmaceuticals       India  
418  pharmaceuticals, Self Made       India  
447        electrical equipment       India  
542                    tractors       India  
828                 diversified       India

In [15]:
# check for NaN values in main dataframe

df.isna().sum()

Name              0
Age              86
Net_Worth         0
Income Source     0
Citizenship       2
dtype: int64

In [16]:
# copy NaN values to a new dataframe 'na' incase of future needs
na = df[df.isna().any(axis=1)]

# drop NaN values from df
df.dropna(inplace=True)

In [17]:
# again add back the invalid data to the main dataframe which is required for further questions

# df must be used for 'Age' related questions cause it doesn't contains NaN values in 'Age' column
df = pd.concat([df, self_made])

# dfs can be used for non 'Age' related questions
dfs = pd.concat([df, self_made, india, na])

# Q4: What is the distribution of the age of billionaires in the dataset? (Medium)
* PreProcess the Age column first either drop all the values that are not required or convert them to NAN {This is a very important step will be used in further questions as well} 

In [18]:
# change dtype of 'Age' column to 'int'
df['Age'] = df['Age'].astype(int)

In [19]:
# plotly distribution charts (hover over for interactive response)

fig = make_subplots(2, 1)

fig.add_trace(go.Histogram(x=df['Age'], name='Age Histogram'), 1, 1)
fig.add_trace(go.Box(x=df['Age'], name='Age Boxplot', boxmean='sd'), 2, 1)

iplot(fig)

### The distribution of the age of billionaires in the dataset

#### - Ages range from 19 to 100
#### - Obviously, 19, 25 and 26 are outliers
#### - Inter-Quartile Range(IQR): the middle 50% of data lie between 55 and 74
#### - The most number of billionaires are at the age 56 - 59

# Q5: Who is the 10th Richest Indian Billionaire  (Medium)
* Drop the Duplicate Values using Name then proceed further

In [20]:
# strip values (dfs)

dfs['Name'] = dfs['Name'].str.strip()
dfs['Income Source'] = dfs['Income Source'].str.strip()
dfs['Citizenship'] = dfs['Citizenship'].str.strip()

In [21]:
# drop duplicates with 'Name' column

dfs.drop_duplicates(subset='Name', inplace=True)

In [22]:
# boolean mask 'Citizenship' to India & 'Name' doesn't contains 'family' and take top 10 'Net_Worth'

dfs[(dfs['Citizenship'] == 'India') & ( ~ dfs['Name'].str.contains('family'))].nlargest(10, 'Net_Worth')

Name Age  Net_Worth  \
1177       Mukesh Ambani  65       94.1   
1756        Gautam Adani  60       74.8   
2098          Shiv Nadar  76       28.7   
760     Cyrus Poonawalla  80       24.3   
2259  Radhakishan Damani  67       20.0   
811       Lakshmi Mittal  71       17.9   
114       Savitri Jindal  72       17.2   
806       Dilip Shanghvi  66       15.6   
1345         Kumar Birla  55       15.5   
1242        Sunil Mittal  64       14.7   

                               Income Source Citizenship  
1177                             diversified       India  
1756  commodities, infrastructure, Self Made       India  
2098                              Technology       India  
760                               Healthcare       India  
2259                        Fashion & Retail       India  
811                          Metals & Mining       India  
114                                    steel       India  
806                               Healthcare       India  
1345                             commodities       India  
1242                      telecom, Self Made       India

### **Sunil Mittal** *($14.7 B)* is the 10th Richest Indian Billionaire

# Q6: What is the average net worth of billionaires in the technology industry?
* Only find for billionaires having single Income Source 
* Dont consider billionaires having more than 1 Income Source

In [23]:
# boolean mask 'Income Source' to Technology, take 'Net_Worth' column to find mean

dfs[dfs['Income Source'] == 'Technology']['Net_Worth'].mean()

6.340882352941176

### **$6.34 B** is the average net worth of billionaires in the *technology industry*

# Q7: What is the position of the 10th Richest Billionaire from Tech Industry in the Forbes List (Medium)
* Create an Index showing the rank according to Net Worth (Sort then create index column)
* Then Proceed further with Technology as Income Source

In [24]:
# sort values by DESC of 'Net_Worth' column and drop the previous index

dfs = dfs.sort_values('Net_Worth', ascending=False).reset_index(drop=True)

In [25]:
# boolean mask 'Income Source' to Technology and take the 9th record which is the 10th rank
# and look into the index for the overall position

dfs[dfs['Income Source'] == 'Technology'].iloc[9:10]

Name Age  Net_Worth Income Source Citizenship
34  Ma Huateng  50       37.2    Technology       China

### The position of *the 10th Richest Billionaire from Tech Industry* in the Forbes List is **35**

# Q8: What is the total net worth of the top 100 billionaires in the Forbes dataset, and what percentage of the total net worth of all billionaires does this represent?
* Find Net Worth held by top 100 People
* Then find the percentage of the answer with the total net worth

In [26]:
# sum of top 100 and total 'Net_Worth'

top100 = round(dfs['Net_Worth'][:100].sum(), 2)
total = round(dfs['Net_Worth'].sum(), 2)

In [27]:
top100

4323.0

In [28]:
# percent of top 100 billionaire's 'Net_Worth'

round(top100/total * 100, 2)

33.28

### The total net worth of the top 100 billionaires is **$4323 B** and it represents *33.28%(one third)* of the total net worth of all billionaires

# Q9: What is the average age of billionaires in the Technology industry? 

In [29]:
# strip values (df)

df['Name'] = df['Name'].str.strip()
df['Income Source'] = df['Income Source'].str.strip()
df['Citizenship'] = df['Citizenship'].str.strip()

In [30]:
# boolean mask 'Income Source' to Technology, take 'Age' column to find mean

df[df['Income Source'] == 'Technology']['Age'].mean()

56.16918429003021

### **56** is the average age of billionaires in the *Technology* industry

# Q10: Draw a Graph depicting mean Net Worth of billionaires per Country(Medium)
* For some people Education has been written in Citizenship Column find those and change there citizenship to India
* After doing  it plot the graph

The data is cleaned already

In [31]:
# group 'Citizenship', find mean by 'Net_Worth' and sort values in DESC to better visualize the data

country_networth_avg = dfs.groupby('Citizenship')['Net_Worth'].mean().reset_index(name='mean net worth').sort_values('mean net worth', ascending=False)

In [32]:
# bar graph

fig = px.bar(country_networth_avg, x='Citizenship', y='mean net worth', title='Mean Net Worth of billionaires per Country')

iplot(fig)

# Q11: Top 5 industries that has the most billionaires?

In [33]:
# group 'Income Source', count the size of each Industries and get the top 5

dfs.groupby('Income Source').size().reset_index(name='count').nlargest(5, 'count')

Income Source  count
5   Finance & Investments    392
16             Technology    340
10          Manufacturing    333
4        Fashion & Retail    249
8              Healthcare    213

### *Top 5 industries* that has the most billionaires are **Finance & Investments, Technology, Manufacturing, Fashion & Retail and Healthcare**

# Q12: Create a Violinplot that shows the distribution of Net Worths(Greater than 50B) across different Industries

In [34]:
# Create violin plot

fig = px.violin(dfs[dfs['Net_Worth'] > 50], x='Income Source', y='Net_Worth', box=True, points='all',
                title='Distribution of Net Worths (Greater than 50B) across Industries')

iplot(fig)

# Q13: Create a Heatmap that shows the correlation between age and net worth for billionaires in the technology industry?

In [35]:
# boolean mask 'Income Source' to Technology
tech = df[df['Income Source'] == 'Technology']

# take 'Age' and 'Net_Worth' column
subset = tech[['Age', 'Net_Worth']]

# correlation
corr_matrix = subset.corr()

# heat map
fig = px.imshow(corr_matrix)

fig.update_layout(title="Correlation Heatmap: Age vs Net Worth (Technology Industry)")

iplot(fig)

# Q14: What is the relationship between the age and net worth of billionaires from the top 10 countries? Create a scatter plot matrix to visualize it.

In [36]:
# use top10 variable from Q1 and take only top 10 countries from df

data = df[df['Citizenship'].isin(top10['Citizenship'])]

In [37]:
# Create the scatter plot matrix (click on a label to remove that plot from the graph, its interactive)

fig = px.scatter_matrix(data, dimensions=['Age', 'Net_Worth'], color='Citizenship',
    title='Scatter Plot Matrix: Age Vs. Net Worth of Billionaires (Top 10 Countries)')

iplot(fig)

# Q15: What is the average age of billionaires in the USA? How does it compare to the average age of billionaires in other countries? 
* Find Mean of Age when Citizenship == USA
* Find Mean of Age when Citizenship is not equal to USA(All other countries other than USA)

In [38]:
# boolean mask 'Citizenship' to 'USA' and other countries
# and to find the mean of 'Age' column

usa = df[df['Citizenship'] == 'United States']['Age'].mean()
others = df[df['Citizenship'] != 'USA']['Age'].mean()

In [39]:
print(round(usa, 2))
print(round(others, 2))

66.45
64.39


### The average age of billionaires in the USA is **66.45** and it's slightly higher than the average age of billionaires in other countries *(64.39)*